### PR TITLE
#257: Feat: Added support for HGETALL

### PR DESCRIPTION
**Pull request by raghavbabbar**
https://github.com/DiceDB/dice/issues/257

### CODE REVIEW

File: tests/hgetall_test.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+		},
+	}
+
+	for _, tc := range testCases {
+		for i, cmd := range tc.commands {
+			result := fireCommand(conn, cmd)
+			expectedResults, ok := tc.expected[i].([]string)
+			results, ok2 := result.([]interface{})
+
+			if ok && ok2 && len(results) == len(expectedResults) {
```
Comment (raghavbabbar) [inline tests/hgetall_test.go:39]: did this because results can be unordered.

**Comment by raghavbabbar** [comment]
@JyotinderSingh  Could you please review it

**Comment by raghavbabbar** [comment]
> Some tests are failing, could you please take a look?

On local they are passing, don't know what's going wrong.. let me double check

**Comment by raghavbabbar** [comment]
earlier the test were failing because all test run simuntaniously for same port, and some key clash were there.. leading to invalid results. as of now i have changed the key name for my test.. but i am thinking should there be any other way around. @JyotinderSingh

File: tests/hgetall_test.go [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+				results, ok2 := result.([]interface{})
+
+				if ok && ok2 && len(results) == len(expectedResults) {
+					expectedResultsMap := make(map[string]string)
+					resultsMap := make(map[string]string)
+
+					for i := 0; i < len(results); i += 2 {
+						expectedResultsMap[expectedResults[i]] = expectedResults[i+1]
+						resultsMap[results[i].(string)] = results[i+1].(string)
+					}
```
Comment (raghavbabbar) [inline tests/hgetall_test.go:47]: made map because as results can be in unordered state too.

**Comment by raghavbabbar** [comment]
@JyotinderSingh  please review

File: core/eval.go [hunk 3]
```diff
@@ -2469,6 +2469,34 @@ func evalHSET(args []string, store *Store) []byte {
...
+	var results []string
+
+	if obj != nil {
+		switch currentVal := obj.Value.(type) {
+		case HashMap:
+			hashMap = currentVal
+		default:
+			return diceerrors.NewErrWithMessage(diceerrors.WrongTypeErr)
+		}
+	}
```
Comment (JyotinderSingh) [inline core/eval.go:2491]: Let's use assertTypeEncoding for this check, and please do the same for existing evalHSET method.

File: tests/hgetall_test.go [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+			expected: []interface{}{"OK", "WRONGTYPE Operation against a key holding the wrong kind of value"},
+		},
+		{
+			commands: []string{"HGETALL key_hGetAll03 x", "HGETALL"},
+			expected: []interface{}{"ERR wrong number of arguments for 'hgetall' command",
+				"ERR wrong number of arguments for 'hgetall' command"},
+		},
+	}
+
+	for _, tc := range testCases {
```
Comment (JyotinderSingh) [inline tests/hgetall_test.go:33]: Run cleanup logic using defer to cleanup the keys after the test is done. That may also be the reason why your tests were failing earlier. All integration tests share the same server, therefore each test needs to ensure it cleans up its resources once its done.

**Review by JyotinderSingh — CHANGES_REQUESTED** [review]
Thanks for adding these changes, have left a few comments.

File: core/eval.go (hunk 3 above)
Comment (raghavbabbar) [reply core/eval.go]: Done

File: tests/hgetall_test.go (hunk 4 above)
Comment (raghavbabbar) [reply tests/hgetall_test.go:34]: yes that was exactly the reason, i already mentioned it over here https://github.com/DiceDB/dice/pull/431#issuecomment-2323299209.
As of now I have deleted keys in defer..
but we may need to find better approach.

**Comment by raghavbabbar** [comment]
@JyotinderSingh  please reivew

**Review by JyotinderSingh — APPROVED** [review]
Thanks for adding support for `HGETALL` @raghavbabbar! The changes look good.

**Pull request merged**